feat(semantic): align class scope with typescript

Dunqing opened the pull request:
```ts
class Klass <T>   extends Root       <R>                   {}
//    ^^^^^ ^^^           ^^^^       ^^^                   ^^
//    id type_paramter super_class super_type_parameters  body
```
I reorder fields according to the order above

The class scope is not defined in the spec. But we need to create a scope for `class` to store `TypeParamters`

Boshen commented on tasks/coverage/parser_test262.snap:
```diff
@@ -2,12 +2,186 @@ commit: a1587416
+  × Identifier `A` has already been declared
+    ╭─[built-ins/Function/prototype/toString/class-expression-explicit-ctor.js:10:5]
+  9 │ 
+ 10 │ let A = /* before */class /* a */ A /* b */ extends /* c */ B /* d */ { /* e */ constructor /* f */ ( /* g */ ) /* h */ { /* i */ ; /* j */ } /* k */ m /* l */ ( /* m */ ) /* n */ { /* o */ } /* p */ }/* after */;
```
Oh no!

Dunqing replied:
Don't worry! It has already been fixed in https://github.com/oxc-project/oxc/pull/4141

Boshen replied:
Can we merge these 2 PRs, this is confusing :-/

Dunqing replied:
Done

overlookmotel commented:
> The class scope is not defined in the spec

I think it is - please see #4161.

Dunqing commented:
Blocked on https://github.com/oxc-project/oxc/issues/4200

Dunqing commented:
The regression tests have been fixed in #4156. Should I merge into this PR?

Dunqing reviewed (commented):
Not adding symbols to scope may cause some confusion. For example, in mangler, we need to rename bindings, but we can't find the symbols for function expression and class expression in scope.

 I'll revert this change

Dunqing commented on crates/oxc_semantic/src/builder.rs:
```diff
@@ -1586,6 +1546,20 @@ impl<'a> SemanticBuilder<'a> {
+            AstKind::BindingIdentifier(_) => {
+                if let Some(node) = self.nodes.parent_node(self.current_node_id) {
+                    if let AstKind::Class(class) = node.kind() {
+                        if class.is_expression() {
+                            // We must to bind class expression when enter BindingIdentifier,
+                            // because we should add binding to current scope
+                            let previous_node_id = self.current_node_id;
+                            self.current_node_id = node.id();
+                            class.bind(self);
+                            self.current_node_id = previous_node_id;
+                        }
```
Bind `ClassExpression` in `BindingIdentifier` to add the binding to the correct scope

Dunqing committed:
feat(semantic): align class scope with typescript (#4195)

```ts
class Klass <T>   extends Root       <R>                   {}
//    ^^^^^ ^^^           ^^^^       ^^^                   ^^
//    id type_paramter super_class super_type_parameters  body
```
I reorder fields according to the order above

The class scope is not defined in the spec. But we need to create a scope for `class` to store `TypeParamters`
